src/phantasm-renderer/Context.cc

```diff
@@ -244,42 +250,41 @@ void Context::free_to_cache(const render_target& rt) { freeCachedTarget(rt.info,
 void Context::write_to_buffer_raw(const buffer& buffer, cc::span<std::byte const> data, size_t offset_in_buffer)
 {
     CC_ASSERT(buffer.info.heap == phi::resource_heap::upload && "Attempted to write to non-upload buffer");
-    CC_ASSERT(buffer.info.size_bytes >= data.size() + offset_in_buffer && "Buffer write out of bounds");
+    CC_ASSERT(buffer.info.size_bytes >= data.size_bytes() + offset_in_buffer && "Buffer write out of bounds");
+
+    int const map_begin = int(offset_in_buffer);
+    int const map_end = int(offset_in_buffer + data.size_bytes());
 
-    auto* const map = map_buffer(buffer);
-    std::memcpy(map + offset_in_buffer, data.data(), data.size());
-    unmap_buffer(buffer);
+    std::byte* const map = map_buffer(buffer, map_begin, map_end);
+    std::memcpy(map + offset_in_buffer, data.data(), data.size_bytes());
+    unmap_buffer(buffer, map_begin, map_end);
 }
 
 void Context::read_from_buffer_raw(const buffer& buffer, cc::span<std::byte> out_data, size_t offset_in_buffer)
 {
     CC_ASSERT(buffer.info.heap == phi::resource_heap::readback && "Attempted to read from non-readback buffer");
-    CC_ASSERT(buffer.info.size_bytes >= out_data.size() + offset_in_buffer && "Buffer read out of bounds");
+    CC_ASSERT(buffer.info.size_bytes >= out_data.size_bytes() + offset_in_buffer && "Buffer read out of bounds");
+
+    int const map_begin = int(offset_in_buffer);
+    int const map_end = int(offset_in_buffer + out_data.size_bytes());
 
-    auto* const map = map_buffer(buffer);
-    std::memcpy(out_data.data(), map + offset_in_buffer, out_data.size());
-    unmap_buffer(buffer);
+    std::byte const* const map = map_buffer(buffer, map_begin, map_end);
+    std::memcpy(out_data.data(), map + offset_in_buffer, out_data.size_bytes());
+    unmap_buffer(buffer, map_begin, map_end);
 }
 
 void Context::signal_fence_cpu(const fence& fence, uint64_t new_value) { mBackend->signalFenceCPU(fence.handle, new_value); }
 
 void Context::wait_fence_cpu(const fence& fence, uint64_t wait_value) { mBackend->waitFenceCPU(fence.handle, wait_value); }
 
-void Context::signal_fence_gpu(const fence& fence, uint64_t new_value, phi::queue_type queue)
-{
-    mBackend->signalFenceGPU(fence.handle, new_value, queue);
-}
+uint64_t Context::get_fence_value(const fence& fence) { return mBackend->getFenceValue(fence.handle); }
 
-void Context::wait_fence_gpu(const fence& fence, uint64_t wait_value, phi::queue_type queue)
+std::byte* Context::map_buffer(const buffer& buffer, int invalidate_begin, int invalidate_end)
```

Comment:
could this return a `span<std::byte>` instead?

Reply:
added `map_buffer_as_span` (the map_buffer changes here were non-breaking, and using the pointer directly is often more convenient - for higher level APIs we have write_to_buffer / read_from_buffer)